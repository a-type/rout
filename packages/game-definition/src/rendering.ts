import { GameSessionChatMessage } from '@long-game/common';
import { ComponentType } from 'react';

export type GameRendererVersionDefinition = {
  Client: ComponentType<any>;
  ChatMessage: GameChatMessageRenderer;
};

export type GameRendererModuleDefault = Record<
  string,
  GameRendererVersionDefinition
>;

export type GameRendererModule = {
  default: GameRendererModuleDefault;
};

export type GameChatMessageRenderer =
  ComponentType<GameChatMessageRendererProps>;
export type GameChatMessageRendererProps = {
  message: GameSessionChatMessage;
  previousMessage: GameSessionChatMessage | null;
  nextMessage: GameSessionChatMessage | null;
  compact: boolean;
};
