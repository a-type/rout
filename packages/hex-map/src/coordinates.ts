export type HexCoordinate = [number, number];

// shorthand way to make a coordinate
export function coord(q: number, r: number): HexCoordinate {
  return [q, r];
}

export function coordinatesEqual(a: HexCoordinate, b: HexCoordinate): boolean {
  return a[0] === b[0] && a[1] === b[1];
}

// the 'hidden' third coordinate value; q + r + s = 0 by definition
// so it's always "- q - r"
export function getS(coord: HexCoordinate) {
  return -coord[0] - coord[1];
}

export function fromRS(r: number, s: number): HexCoordinate {
  return [-r - s, r];
}
export function fromQS(q: number, s: number): HexCoordinate {
  return [q, -q - s];
}

export function addCoordinates(
  a: HexCoordinate,
  b: HexCoordinate,
): HexCoordinate {
  return [a[0] + b[0], a[1] + b[1]];
}

export function subtractCoordinates(
  a: HexCoordinate,
  b: HexCoordinate,
): HexCoordinate {
  return addCoordinates(a, [-b[0], -b[1]]);
}

export function multiplyCoordinate(
  coord: HexCoordinate,
  scale: number,
): HexCoordinate {
  return [coord[0] * scale, coord[1] * scale];
}

export function coordinateLength(a: HexCoordinate) {
  return Math.floor((Math.abs(a[0]) + Math.abs(a[1]) + Math.abs(getS(a))) / 2);
}

export function coordinateDistance(a: HexCoordinate, b: HexCoordinate): number {
  return coordinateLength(subtractCoordinates(a, b));
}

const coordinateDirections: HexCoordinate[] = [
  [0, -1],
  [1, -1],
  [1, 0],
  [0, 1],
  [-1, 1],
  [-1, 0],
];
/**
 * Gets a directional hex vector by index.
 *
 *       0
 * 5    ---    1
 *    /  q  \
 *    \s   r/
 * 4    ---    2
 *       3
 */
export function getDirection(dir: number): HexCoordinate {
  return coordinateDirections[(6 + (dir % 6)) % 6];
}

export function getNeighbor(coord: HexCoordinate, dir: number) {
  const direction = getDirection(dir);
  return addCoordinates(coord, direction);
}

function lerp(a: number, b: number, t: number): number {
  return a + (b - a) * t;
}

export function lerpCoordinates(
  start: HexCoordinate,
  end: HexCoordinate,
  t: number,
): HexCoordinate {
  return [lerp(start[0], end[0], t), lerp(start[1], end[1], t)];
}

export function rotateCoordinate(
  coord: HexCoordinate,
  /** Only rotations in increments of 60 deg (Pi / 3) are supported */
  turnsOf60Degrees: number,
  origin: HexCoordinate = [0, 0],
) {
  const direction = Math.sign(turnsOf60Degrees);
  for (let i = 0; i < Math.abs(turnsOf60Degrees); i++) {
    coord = rotateCoordinateOnce(coord, direction, origin);
  }
  return coord;
}

function rotateCoordinateOnce(
  coord: HexCoordinate,
  direction: number,
  origin: HexCoordinate = [0, 0],
): HexCoordinate {
  const adjusted = subtractCoordinates(coord, origin);
  const s = getS(adjusted);
  if (direction < 0) {
    return addCoordinates(origin, [-s, -adjusted[0]]);
  } else {
    return addCoordinates(origin, [-adjusted[1], -s]);
  }
}
