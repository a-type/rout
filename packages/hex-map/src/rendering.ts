import { getS, HexCoordinate } from './coordinates.js';

// these are 2x2 matrices, but expressed as single array for compactness
type OrientationMatrix = [number, number, number, number];

const pointyForwardMatrix: OrientationMatrix = [
  Math.sqrt(3),
  Math.sqrt(3) / 2,
  0,
  3 / 2,
];
const pointyInverseMatrix: OrientationMatrix = [
  Math.sqrt(3) / 3,
  -1 / 3,
  0,
  2 / 3,
];
const pointyInitialAngleRad = Math.PI / 6; // 30 deg

const flatForwardMatrix: OrientationMatrix = [
  3 / 2,
  0,
  Math.sqrt(3) / 2,
  Math.sqrt(3),
];
const flatInverseMatrix: OrientationMatrix = [
  2 / 3,
  0,
  -1 / 3,
  Math.sqrt(3) / 3,
];
const flatInitialAngleRad = 0;

type LayoutDetails = {
  forward: OrientationMatrix;
  inverse: OrientationMatrix;
  initialAngleRad: number;
};

const layouts: Record<'pointy' | 'flat', LayoutDetails> = {
  pointy: {
    forward: pointyForwardMatrix,
    inverse: pointyInverseMatrix,
    initialAngleRad: pointyInitialAngleRad,
  },
  flat: {
    forward: flatForwardMatrix,
    inverse: flatInverseMatrix,
    initialAngleRad: flatInitialAngleRad,
  },
};

export interface HexLayoutContext {
  orientation: 'pointy' | 'flat';
  size: [number, number]; // width, height in pixels of hex tiles. set to same for uniform hex, can be different for squashed (like pixel art)
  origin?: [number, number]; // origin in pixels of the hex grid, if not 0,0
}

function applyLayoutMatrix(coord: HexCoordinate, matrix: OrientationMatrix) {
  const x = matrix[0] * coord[0] + matrix[1] * coord[1];
  const y = matrix[2] * coord[0] + matrix[3] * coord[1];
  return [x, y];
}

export function coordinateToScreenCenter(
  coord: HexCoordinate,
  ctx: HexLayoutContext,
) {
  const layout = layouts.pointy;
  const [x, y] = applyLayoutMatrix(coord, layout.forward);
  return [
    x * ctx.size[0] + (ctx.origin?.[0] || 0),
    y * ctx.size[1] + (ctx.origin?.[1] || 0),
  ] as [number, number];
}

/**
 * Returns a fractional hex coordinate. Use roundFractionalCoordinate to get the true hex coordinate.
 */
export function screenToCoordinate(
  pixel: [number, number],
  ctx: HexLayoutContext,
): HexCoordinate {
  const layout = layouts[ctx.orientation];
  const [x, y] = applyLayoutMatrix(pixel, layout.inverse);
  return [
    x / ctx.size[0] - (ctx.origin?.[0] || 0),
    y / ctx.size[1] - (ctx.origin?.[1] || 0),
  ] as [number, number];
}

export function roundFractionalCoordinate(
  fractional: HexCoordinate,
): HexCoordinate {
  const originalS = getS(fractional);
  const rounded = fractional.map(Math.round) as HexCoordinate;
  let s = getS(rounded);
  let [q, r] = rounded;

  const qDiff = Math.abs(q - fractional[0]);
  const rDiff = Math.abs(r - fractional[1]);
  const sDiff = Math.abs(s - originalS);

  if (qDiff > rDiff && qDiff > sDiff) {
    q = -r - s;
  } else if (rDiff > sDiff) {
    r = -q - s;
  } else {
    s = -q - r;
  }

  return [r, q];
}

export function cornerOffset(
  corner: number,
  ctx: HexLayoutContext,
): [number, number] {
  const layout = layouts[ctx.orientation];
  const angle = layout.initialAngleRad + (corner * Math.PI) / 3;
  return [Math.cos(angle) * ctx.size[0], Math.sin(angle) * ctx.size[1]];
}

export function getHexCorners(
  coord: HexCoordinate,
  ctx: HexLayoutContext,
): [number, number][] {
  return Array.from({ length: 6 }, (_, i) => cornerOffset(i, ctx)).map(
    (offset) => [offset[0] + coord[0], offset[1] + coord[1]],
  );
}
