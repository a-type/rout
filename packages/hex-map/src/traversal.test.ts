import { describe, expect, it } from 'vitest';
import { coordinatesEqual, HexCoordinate } from './coordinates.js';
import { HexLayoutContext } from './rendering.js';
import {
  collectTraversals,
  diamondTraverser,
  hexagonTraverser,
  lineTraverser,
  pathfind,
  rectangleTraverser,
  ringTraverser,
  spiralTraverser,
  triangleTraverser,
} from './traversal.js';

describe('map traversers', () => {
  describe.each(['pointy', 'flat'] as const)(
    'layout: %s top',
    (orientation) => {
      const ctx: HexLayoutContext = {
        orientation,
        size: [100, 100],
      };

      it('traverses diamonds', () => {
        expect(collectTraversals(diamondTraverser(ctx, 2))).toMatchSnapshot();
      });
      it('traverses triangles', () => {
        expect(
          collectTraversals(triangleTraverser(ctx, 6, 0)),
        ).toMatchSnapshot();
      });
      it('traverses hexagons', () => {
        expect(collectTraversals(hexagonTraverser(ctx, 3))).toMatchSnapshot();
      });
      it('traverses rectangles', () => {
        expect(
          collectTraversals(rectangleTraverser(ctx, 7, 5)),
        ).toMatchSnapshot();
      });
      it('traverses rings', () => {
        expect(collectTraversals(ringTraverser(ctx, 3))).toMatchSnapshot();
      });
      it('traverses spirals', () => {
        expect(collectTraversals(spiralTraverser(ctx, 3))).toMatchSnapshot();
      });
    },
  );

  it('traverses lines (orientation independent)', () => {
    expect(collectTraversals(lineTraverser([-2, 2], [2, 2]))).toMatchSnapshot();
    expect(collectTraversals(lineTraverser([0, 0], [4, 1]))).toMatchSnapshot();
  });

  it('traverses a ring of radius 0', () => {
    expect(
      collectTraversals(
        ringTraverser({ orientation: 'pointy', size: [100, 100] }, 0),
      ),
    ).toEqual([[0, 0]]);
  });
});

describe('pathfinding', () => {
  it('can find a path from start to end', () => {
    const start: HexCoordinate = [0, 0];
    const end: HexCoordinate = [3, -3];
    const obstacleTest = (coord: HexCoordinate) => {
      if (coordinatesEqual(coord, [1, -1])) return true;
      return false;
    };

    const path = pathfind(start, end, obstacleTest, 10);
    expect(path).toEqual([
      [0, 0],
      [0, -1],
      [1, -2],
      [2, -3],
      [3, -3],
    ]);
  });

  it('detects impossible paths', () => {
    const start: HexCoordinate = [0, 0];
    const end: HexCoordinate = [3, -3];
    const obstacleTest = (coord: HexCoordinate) => {
      // wall it in
      if (
        (
          [
            [2, -3],
            [3, -4],
            [4, -4],
            [4, -3],
            [3, -2],
            [2, -2],
          ] as HexCoordinate[]
        ).some((c) => coordinatesEqual(coord, c))
      )
        return true;
      return false;
    };

    const path = pathfind(start, end, obstacleTest, 10);
    expect(path).toBe(null);
  });

  it('detects paths too long to reach due to straight length', () => {
    const start: HexCoordinate = [0, 0];
    const end: HexCoordinate = [3, -3];
    const obstacleTest = () => false;
    expect(pathfind(start, end, obstacleTest, 3)).toBe(null);
    // reachable with greater allowed length
    expect(pathfind(start, end, obstacleTest, 4)).not.toBe(null);
  });

  it('detects paths too long to reach due to obstacles', () => {
    const start: HexCoordinate = [0, 0];
    const end: HexCoordinate = [3, -3];
    const obstacleTest = (coord: HexCoordinate) => {
      // ring around the origin, with only the furthest away
      // spot open
      return (
        [
          [-1, 0],
          [0, -1],
          [1, -1],
          [1, 0],
          [0, 1],
        ] as HexCoordinate[]
      ).some((c) => coordinatesEqual(coord, c));
    };

    expect(pathfind(start, end, obstacleTest, 8)).toBe(null);
    // reachable with more length
    expect(pathfind(start, end, obstacleTest, 9)).toMatchInlineSnapshot(`
      [
        [
          0,
          0,
        ],
        [
          -1,
          1,
        ],
        [
          -1,
          2,
        ],
        [
          0,
          2,
        ],
        [
          1,
          1,
        ],
        [
          2,
          0,
        ],
        [
          2,
          -1,
        ],
        [
          2,
          -2,
        ],
        [
          3,
          -3,
        ],
      ]
    `);
  });
});
