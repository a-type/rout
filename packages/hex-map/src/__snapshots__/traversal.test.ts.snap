// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`map traversers > layout: flat top > traverses diamonds 1`] = `
[
  [
    4,
    -2,
  ],
  [
    3,
    -2,
  ],
  [
    2,
    -2,
  ],
  [
    1,
    -2,
  ],
  [
    0,
    -2,
  ],
  [
    3,
    -1,
  ],
  [
    2,
    -1,
  ],
  [
    1,
    -1,
  ],
  [
    0,
    -1,
  ],
  [
    -1,
    -1,
  ],
  [
    2,
    0,
  ],
  [
    1,
    0,
  ],
  [
    0,
    0,
  ],
  [
    -1,
    0,
  ],
  [
    -2,
    0,
  ],
  [
    1,
    1,
  ],
  [
    0,
    1,
  ],
  [
    -1,
    1,
  ],
  [
    -2,
    1,
  ],
  [
    -3,
    1,
  ],
  [
    0,
    2,
  ],
  [
    -1,
    2,
  ],
  [
    -2,
    2,
  ],
  [
    -3,
    2,
  ],
  [
    -4,
    2,
  ],
]
`;

exports[`map traversers > layout: flat top > traverses hexagons 1`] = `
[
  [
    -3,
    0,
  ],
  [
    -3,
    1,
  ],
  [
    -3,
    2,
  ],
  [
    -3,
    3,
  ],
  [
    -2,
    -1,
  ],
  [
    -2,
    0,
  ],
  [
    -2,
    1,
  ],
  [
    -2,
    2,
  ],
  [
    -2,
    3,
  ],
  [
    -1,
    -2,
  ],
  [
    -1,
    -1,
  ],
  [
    -1,
    0,
  ],
  [
    -1,
    1,
  ],
  [
    -1,
    2,
  ],
  [
    -1,
    3,
  ],
  [
    0,
    -3,
  ],
  [
    0,
    -2,
  ],
  [
    0,
    -1,
  ],
  [
    0,
    0,
  ],
  [
    0,
    1,
  ],
  [
    0,
    2,
  ],
  [
    0,
    3,
  ],
  [
    1,
    -3,
  ],
  [
    1,
    -2,
  ],
  [
    1,
    -1,
  ],
  [
    1,
    0,
  ],
  [
    1,
    1,
  ],
  [
    1,
    2,
  ],
  [
    2,
    -3,
  ],
  [
    2,
    -2,
  ],
  [
    2,
    -1,
  ],
  [
    2,
    0,
  ],
  [
    2,
    1,
  ],
  [
    3,
    -3,
  ],
  [
    3,
    -2,
  ],
  [
    3,
    -1,
  ],
  [
    3,
    0,
  ],
]
`;

exports[`map traversers > layout: flat top > traverses rectangles 1`] = `
[
  [
    0,
    0,
  ],
  [
    0,
    1,
  ],
  [
    0,
    2,
  ],
  [
    0,
    3,
  ],
  [
    0,
    4,
  ],
  [
    0,
    5,
  ],
  [
    1,
    0,
  ],
  [
    1,
    1,
  ],
  [
    1,
    2,
  ],
  [
    1,
    3,
  ],
  [
    1,
    4,
  ],
  [
    1,
    5,
  ],
  [
    2,
    -1,
  ],
  [
    2,
    0,
  ],
  [
    2,
    1,
  ],
  [
    2,
    2,
  ],
  [
    2,
    3,
  ],
  [
    2,
    4,
  ],
  [
    3,
    -1,
  ],
  [
    3,
    0,
  ],
  [
    3,
    1,
  ],
  [
    3,
    2,
  ],
  [
    3,
    3,
  ],
  [
    3,
    4,
  ],
  [
    4,
    -2,
  ],
  [
    4,
    -1,
  ],
  [
    4,
    0,
  ],
  [
    4,
    1,
  ],
  [
    4,
    2,
  ],
  [
    4,
    3,
  ],
  [
    5,
    -2,
  ],
  [
    5,
    -1,
  ],
  [
    5,
    0,
  ],
  [
    5,
    1,
  ],
  [
    5,
    2,
  ],
  [
    5,
    3,
  ],
  [
    6,
    -3,
  ],
  [
    6,
    -2,
  ],
  [
    6,
    -1,
  ],
  [
    6,
    0,
  ],
  [
    6,
    1,
  ],
  [
    6,
    2,
  ],
  [
    7,
    -3,
  ],
  [
    7,
    -2,
  ],
  [
    7,
    -1,
  ],
  [
    7,
    0,
  ],
  [
    7,
    1,
  ],
  [
    7,
    2,
  ],
]
`;

exports[`map traversers > layout: flat top > traverses rings 1`] = `
[
  [
    -3,
    3,
  ],
  [
    -3,
    2,
  ],
  [
    -3,
    1,
  ],
  [
    -3,
    0,
  ],
  [
    -2,
    -1,
  ],
  [
    -1,
    -2,
  ],
  [
    0,
    -3,
  ],
  [
    1,
    -3,
  ],
  [
    2,
    -3,
  ],
  [
    3,
    -3,
  ],
  [
    3,
    -2,
  ],
  [
    3,
    -1,
  ],
  [
    3,
    0,
  ],
  [
    2,
    1,
  ],
  [
    1,
    2,
  ],
  [
    0,
    3,
  ],
  [
    -1,
    3,
  ],
  [
    -2,
    3,
  ],
]
`;

exports[`map traversers > layout: flat top > traverses spirals 1`] = `
[
  [
    0,
    0,
  ],
  [
    -1,
    1,
  ],
  [
    -1,
    0,
  ],
  [
    0,
    -1,
  ],
  [
    1,
    -1,
  ],
  [
    1,
    0,
  ],
  [
    0,
    1,
  ],
  [
    -2,
    2,
  ],
  [
    -2,
    1,
  ],
  [
    -2,
    0,
  ],
  [
    -1,
    -1,
  ],
  [
    0,
    -2,
  ],
  [
    1,
    -2,
  ],
  [
    2,
    -2,
  ],
  [
    2,
    -1,
  ],
  [
    2,
    0,
  ],
  [
    1,
    1,
  ],
  [
    0,
    2,
  ],
  [
    -1,
    2,
  ],
  [
    -3,
    3,
  ],
  [
    -3,
    2,
  ],
  [
    -3,
    1,
  ],
  [
    -3,
    0,
  ],
  [
    -2,
    -1,
  ],
  [
    -1,
    -2,
  ],
  [
    0,
    -3,
  ],
  [
    1,
    -3,
  ],
  [
    2,
    -3,
  ],
  [
    3,
    -3,
  ],
  [
    3,
    -2,
  ],
  [
    3,
    -1,
  ],
  [
    3,
    0,
  ],
  [
    2,
    1,
  ],
  [
    1,
    2,
  ],
  [
    0,
    3,
  ],
  [
    -1,
    3,
  ],
  [
    -2,
    3,
  ],
]
`;

exports[`map traversers > layout: flat top > traverses triangles 1`] = `
[
  [
    0,
    0,
  ],
  [
    0,
    1,
  ],
  [
    0,
    2,
  ],
  [
    0,
    3,
  ],
  [
    0,
    4,
  ],
  [
    0,
    5,
  ],
  [
    1,
    0,
  ],
  [
    1,
    1,
  ],
  [
    1,
    2,
  ],
  [
    1,
    3,
  ],
  [
    1,
    4,
  ],
  [
    2,
    0,
  ],
  [
    2,
    1,
  ],
  [
    2,
    2,
  ],
  [
    2,
    3,
  ],
  [
    3,
    0,
  ],
  [
    3,
    1,
  ],
  [
    3,
    2,
  ],
  [
    4,
    0,
  ],
  [
    4,
    1,
  ],
  [
    5,
    0,
  ],
]
`;

exports[`map traversers > layout: pointy top > traverses diamonds 1`] = `
[
  [
    -2,
    4,
  ],
  [
    -1,
    3,
  ],
  [
    0,
    2,
  ],
  [
    1,
    1,
  ],
  [
    2,
    0,
  ],
  [
    -2,
    3,
  ],
  [
    -1,
    2,
  ],
  [
    0,
    1,
  ],
  [
    1,
    0,
  ],
  [
    2,
    -1,
  ],
  [
    -2,
    2,
  ],
  [
    -1,
    1,
  ],
  [
    0,
    0,
  ],
  [
    1,
    -1,
  ],
  [
    2,
    -2,
  ],
  [
    -2,
    1,
  ],
  [
    -1,
    0,
  ],
  [
    0,
    -1,
  ],
  [
    1,
    -2,
  ],
  [
    2,
    -3,
  ],
  [
    -2,
    0,
  ],
  [
    -1,
    -1,
  ],
  [
    0,
    -2,
  ],
  [
    1,
    -3,
  ],
  [
    2,
    -4,
  ],
]
`;

exports[`map traversers > layout: pointy top > traverses hexagons 1`] = `
[
  [
    -3,
    0,
  ],
  [
    -3,
    1,
  ],
  [
    -3,
    2,
  ],
  [
    -3,
    3,
  ],
  [
    -2,
    -1,
  ],
  [
    -2,
    0,
  ],
  [
    -2,
    1,
  ],
  [
    -2,
    2,
  ],
  [
    -2,
    3,
  ],
  [
    -1,
    -2,
  ],
  [
    -1,
    -1,
  ],
  [
    -1,
    0,
  ],
  [
    -1,
    1,
  ],
  [
    -1,
    2,
  ],
  [
    -1,
    3,
  ],
  [
    0,
    -3,
  ],
  [
    0,
    -2,
  ],
  [
    0,
    -1,
  ],
  [
    0,
    0,
  ],
  [
    0,
    1,
  ],
  [
    0,
    2,
  ],
  [
    0,
    3,
  ],
  [
    1,
    -3,
  ],
  [
    1,
    -2,
  ],
  [
    1,
    -1,
  ],
  [
    1,
    0,
  ],
  [
    1,
    1,
  ],
  [
    1,
    2,
  ],
  [
    2,
    -3,
  ],
  [
    2,
    -2,
  ],
  [
    2,
    -1,
  ],
  [
    2,
    0,
  ],
  [
    2,
    1,
  ],
  [
    3,
    -3,
  ],
  [
    3,
    -2,
  ],
  [
    3,
    -1,
  ],
  [
    3,
    0,
  ],
]
`;

exports[`map traversers > layout: pointy top > traverses rectangles 1`] = `
[
  [
    0,
    0,
  ],
  [
    1,
    0,
  ],
  [
    2,
    0,
  ],
  [
    3,
    0,
  ],
  [
    4,
    0,
  ],
  [
    5,
    0,
  ],
  [
    6,
    0,
  ],
  [
    0,
    1,
  ],
  [
    1,
    1,
  ],
  [
    2,
    1,
  ],
  [
    3,
    1,
  ],
  [
    4,
    1,
  ],
  [
    5,
    1,
  ],
  [
    6,
    1,
  ],
  [
    -1,
    2,
  ],
  [
    0,
    2,
  ],
  [
    1,
    2,
  ],
  [
    2,
    2,
  ],
  [
    3,
    2,
  ],
  [
    4,
    2,
  ],
  [
    5,
    2,
  ],
  [
    -1,
    3,
  ],
  [
    0,
    3,
  ],
  [
    1,
    3,
  ],
  [
    2,
    3,
  ],
  [
    3,
    3,
  ],
  [
    4,
    3,
  ],
  [
    5,
    3,
  ],
  [
    -2,
    4,
  ],
  [
    -1,
    4,
  ],
  [
    0,
    4,
  ],
  [
    1,
    4,
  ],
  [
    2,
    4,
  ],
  [
    3,
    4,
  ],
  [
    4,
    4,
  ],
]
`;

exports[`map traversers > layout: pointy top > traverses rings 1`] = `
[
  [
    -3,
    3,
  ],
  [
    -3,
    2,
  ],
  [
    -3,
    1,
  ],
  [
    -3,
    0,
  ],
  [
    -2,
    -1,
  ],
  [
    -1,
    -2,
  ],
  [
    0,
    -3,
  ],
  [
    1,
    -3,
  ],
  [
    2,
    -3,
  ],
  [
    3,
    -3,
  ],
  [
    3,
    -2,
  ],
  [
    3,
    -1,
  ],
  [
    3,
    0,
  ],
  [
    2,
    1,
  ],
  [
    1,
    2,
  ],
  [
    0,
    3,
  ],
  [
    -1,
    3,
  ],
  [
    -2,
    3,
  ],
]
`;

exports[`map traversers > layout: pointy top > traverses spirals 1`] = `
[
  [
    0,
    0,
  ],
  [
    -1,
    1,
  ],
  [
    -1,
    0,
  ],
  [
    0,
    -1,
  ],
  [
    1,
    -1,
  ],
  [
    1,
    0,
  ],
  [
    0,
    1,
  ],
  [
    -2,
    2,
  ],
  [
    -2,
    1,
  ],
  [
    -2,
    0,
  ],
  [
    -1,
    -1,
  ],
  [
    0,
    -2,
  ],
  [
    1,
    -2,
  ],
  [
    2,
    -2,
  ],
  [
    2,
    -1,
  ],
  [
    2,
    0,
  ],
  [
    1,
    1,
  ],
  [
    0,
    2,
  ],
  [
    -1,
    2,
  ],
  [
    -3,
    3,
  ],
  [
    -3,
    2,
  ],
  [
    -3,
    1,
  ],
  [
    -3,
    0,
  ],
  [
    -2,
    -1,
  ],
  [
    -1,
    -2,
  ],
  [
    0,
    -3,
  ],
  [
    1,
    -3,
  ],
  [
    2,
    -3,
  ],
  [
    3,
    -3,
  ],
  [
    3,
    -2,
  ],
  [
    3,
    -1,
  ],
  [
    3,
    0,
  ],
  [
    2,
    1,
  ],
  [
    1,
    2,
  ],
  [
    0,
    3,
  ],
  [
    -1,
    3,
  ],
  [
    -2,
    3,
  ],
]
`;

exports[`map traversers > layout: pointy top > traverses triangles 1`] = `
[
  [
    0,
    0,
  ],
  [
    0,
    1,
  ],
  [
    0,
    2,
  ],
  [
    0,
    3,
  ],
  [
    0,
    4,
  ],
  [
    0,
    5,
  ],
  [
    1,
    0,
  ],
  [
    1,
    1,
  ],
  [
    1,
    2,
  ],
  [
    1,
    3,
  ],
  [
    1,
    4,
  ],
  [
    2,
    0,
  ],
  [
    2,
    1,
  ],
  [
    2,
    2,
  ],
  [
    2,
    3,
  ],
  [
    3,
    0,
  ],
  [
    3,
    1,
  ],
  [
    3,
    2,
  ],
  [
    4,
    0,
  ],
  [
    4,
    1,
  ],
  [
    5,
    0,
  ],
]
`;

exports[`map traversers > traverses lines (orientation independent) 1`] = `
[
  [
    2,
    -2,
  ],
  [
    2,
    -1,
  ],
  [
    2,
    0,
  ],
  [
    2,
    1,
  ],
  [
    2,
    2,
  ],
]
`;

exports[`map traversers > traverses lines (orientation independent) 2`] = `
[
  [
    0,
    0,
  ],
  [
    0,
    1,
  ],
  [
    0,
    2,
  ],
  [
    1,
    2,
  ],
  [
    1,
    3,
  ],
  [
    1,
    4,
  ],
]
`;
